<a href="https://colab.research.google.com/github/moisesrmz/1erProyecto/blob/main/Actividad2analisisPandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC4029 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 2**
Pandas para el análisis de datos en Python

---

*   NOMBRE: Moises Arturo Ramirez Ruiz
*   MATRÍCULA: A01795425


---

En esta actividad usarás el archivo de datos `LaqnData.csv`. Cada fila de este conjunto de datos muestra un registro de medición, por hora, de uno de los siguientes cinco contaminantes del aire:

*   `NO` (óxido nítrico)
*   `NO2` (dióxido de nitrógeno)
*   `NOX` (óxidos nitrosos)
*   `PM10` (partículas suspendidas con un diámetro aerodinámico igual o menor a 10 micrómetros)
*   `PM2.5` (partículas suspendidas con un diámetro aerodinámico igual o menor a 2.5 micrómetros)

Los datos se recopilaron en una ubicación de Londres durante todo el año 2017.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


1.	Descarga el archivo: `LaqnData.csv` y guarda, en un dataframe (`air_df`), todos sus registros.
*   Observa la estructura y contenido del dataframe con los atributos y métodos estudiados (`shape`, `columns`, `head()`, `tail()`, `dtypes`)
*   Calcula el porcentaje de valores faltantes por columna.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import pandas as pd
air_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/LaqnData.csv')
air_df

Site Species   ReadingDateTime  Value   Units Provisional or Ratified
0      CT3      NO  01/01/2017 00:00    3.5  ug m-3                       R
1      CT3      NO  01/01/2017 01:00    3.6  ug m-3                       R
2      CT3      NO  01/01/2017 02:00    2.2  ug m-3                       R
3      CT3      NO  01/01/2017 03:00    2.1  ug m-3                       R
4      CT3      NO  01/01/2017 04:00    3.3  ug m-3                       R
...    ...     ...               ...    ...     ...                     ...
43795  CT3   PM2.5  31/12/2017 19:00   -2.0  ug m-3                       R
43796  CT3   PM2.5  31/12/2017 20:00    6.0  ug m-3                       R
43797  CT3   PM2.5  31/12/2017 21:00    5.0  ug m-3                       R
43798  CT3   PM2.5  31/12/2017 22:00    5.0  ug m-3                       R
43799  CT3   PM2.5  31/12/2017 23:00    6.0  ug m-3                       R

[43800 rows x 6 columns]

In [4]:
air_df.shape

(43800, 6)

In [5]:
air_df.columns

Index(['Site', 'Species', 'ReadingDateTime', 'Value', 'Units',
       'Provisional or Ratified'],
      dtype='object')

In [6]:
air_df.head()

Site Species   ReadingDateTime  Value   Units Provisional or Ratified
0  CT3      NO  01/01/2017 00:00    3.5  ug m-3                       R
1  CT3      NO  01/01/2017 01:00    3.6  ug m-3                       R
2  CT3      NO  01/01/2017 02:00    2.2  ug m-3                       R
3  CT3      NO  01/01/2017 03:00    2.1  ug m-3                       R
4  CT3      NO  01/01/2017 04:00    3.3  ug m-3                       R

In [7]:
air_df.tail()

Site Species   ReadingDateTime  Value   Units Provisional or Ratified
43795  CT3   PM2.5  31/12/2017 19:00   -2.0  ug m-3                       R
43796  CT3   PM2.5  31/12/2017 20:00    6.0  ug m-3                       R
43797  CT3   PM2.5  31/12/2017 21:00    5.0  ug m-3                       R
43798  CT3   PM2.5  31/12/2017 22:00    5.0  ug m-3                       R
43799  CT3   PM2.5  31/12/2017 23:00    6.0  ug m-3                       R

In [8]:
air_df.dtypes

Site                        object
Species                     object
ReadingDateTime             object
Value                      float64
Units                       object
Provisional or Ratified     object
dtype: object

2.	Obtén la cantidad de valores únicos por columna (`nunique()`) para responder:


*   ¿Cuántas mediciones se hicieron? 43800
*   ¿Cuántos contaminantes del aire (`Species`) se analizaron? 5
*   ¿Qué columnas no agregan ningún valor informativo porque el mismo valor en todo el conjunto de datos? Site y Provisional or Ratified Elimínalas del dataframe.
*   La columna `Units` tampoco posee valor informativo. Verifícalo con la función `unique()` y elimínala del dataframe.

In [9]:
air_df.nunique()

Site                          1
Species                       5
ReadingDateTime            8760
Value                      1847
Units                         3
Provisional or Ratified       1
dtype: int64

In [10]:
air_df.shape

(43800, 6)

In [11]:
del air_df['Site']

In [12]:
del air_df['Provisional or Ratified']

In [34]:
air_df['Units'].unique()

KeyError: ignored

In [14]:
del air_df["Units"]

3.	Realiza el conteo de valores por categoría (`value_counts()`) para la columna `Species` y determina si las lecturas de cada contaminante coincide con el total de mediciones. 8760+8760+8760+8760+8760=43800

In [15]:
air_df['Species'].value_counts()

NO       8760
NO2      8760
NOX      8760
PM10     8760
PM2.5    8760
Name: Species, dtype: int64

4.	Utiliza la función `groupby()` para determinar el promedio del valor por contaminante.

In [16]:
air_df.groupby('Species').mean(numeric_only=True)

Value
Species           
NO       15.045115
NO2      38.010185
NOX      61.078661
PM10     22.551704
PM2.5    14.999831

5.	El conjunto de datos está en un formato largo. Aplique la función apropiada para cambiarlo al formato ancho (`ReadingDateTime` como índice y cada contaminante en una columna). Nombre el dataframe resultante como `pvt_df`.

In [17]:
air_df.set_index('ReadingDateTime', inplace=True)

In [35]:
pvt_df=air_df.T

In [38]:
pvt_df

ReadingDateTime 01/01/2017 00:00 01/01/2017 01:00 01/01/2017 02:00  \
Species                       NO               NO               NO   
Value                        3.5              3.6              2.2   

ReadingDateTime 01/01/2017 03:00 01/01/2017 04:00 01/01/2017 05:00  \
Species                       NO               NO               NO   
Value                        2.1              3.3              2.1   

ReadingDateTime 01/01/2017 06:00 01/01/2017 07:00 01/01/2017 08:00  \
Species                       NO               NO               NO   
Value                        4.4              5.1              7.4   

ReadingDateTime 01/01/2017 09:00  ... 31/12/2017 14:00 31/12/2017 15:00  \
Species                       NO  ...            PM2.5            PM2.5   
Value                        6.5  ...              8.0              2.0   

ReadingDateTime 31/12/2017 16:00 31/12/2017 17:00 31/12/2017 18:00  \
Species                    PM2.5            PM2.5            PM2.5   
Value                        2.0              4.0              5.0   

ReadingDateTime 31/12/2017 19:00 31/12/2017 20:00 31/12/2017 21:00  \
Species                    PM2.5            PM2.5            PM2.5   
Value                       -2.0              6.0              5.0   

ReadingDateTime 31/12/2017 22:00 31/12/2017 23:00  
Species                    PM2.5            PM2.5  
Value                        5.0              6.0  

[2 rows x 43800 columns]

6.	A través de la función `describe()` responde:

*   ¿Cuál es el mayor valor de `NO2` registrado? 120.2 ¿A qué día pertenece? 24-ene
*   ¿Cuál es el menor valor de `PM10` registrado?-5.6¿A qué día pertenece?12-sep
*   ¿Cuál es la mediana del `NO`? 7.1 ¿Cómo se interpreta? es el valor que ocupa el lugar central de todos los datos cuando éstos están ordenados de menor a mayor.
*   ¿Cuál es el primer cuartil de `PM2.5`? 7.0 ¿Qué significa? Que el 25% de los datos estan por debajo de este valor.

In [39]:
pvt_df.describe(include='object')

ReadingDateTime 01/01/2017 00:00 01/01/2017 01:00 01/01/2017 02:00  \
count                          2                2                2   
unique                         2                2                2   
top                           NO               NO               NO   
freq                           1                1                1   

ReadingDateTime 01/01/2017 03:00 01/01/2017 04:00 01/01/2017 05:00  \
count                          2                2                2   
unique                         2                2                2   
top                           NO               NO               NO   
freq                           1                1                1   

ReadingDateTime 01/01/2017 06:00 01/01/2017 07:00 01/01/2017 08:00  \
count                          2                2                2   
unique                         2                2                2   
top                           NO               NO               NO   
freq                           1                1                1   

ReadingDateTime 01/01/2017 09:00  ... 31/12/2017 14:00 31/12/2017 15:00  \
count                          2  ...                2                2   
unique                         2  ...                2                2   
top                           NO  ...            PM2.5            PM2.5   
freq                           1  ...                1                1   

ReadingDateTime 31/12/2017 16:00 31/12/2017 17:00 31/12/2017 18:00  \
count                          2                2                2   
unique                         2                2                2   
top                        PM2.5            PM2.5            PM2.5   
freq                           1                1                1   

ReadingDateTime 31/12/2017 19:00 31/12/2017 20:00 31/12/2017 21:00  \
count                          2                2                2   
unique                         2                2                2   
top                        PM2.5            PM2.5            PM2.5   
freq                           1                1                1   

ReadingDateTime 31/12/2017 22:00 31/12/2017 23:00  
count                          2                2  
unique                         2                2  
top                        PM2.5            PM2.5  
freq                           1                1  

[4 rows x 43800 columns]

In [21]:
pvt_df.groupby('Species').max(numeric_only=True)

Value
Species           
NO       401.79999
NO2      120.20000
NOX      734.29999
PM10     633.09998
PM2.5    128.00000

In [22]:
pvt_df[(pvt_df['Species'] =='NO2') & (pvt_df["Value"] == 120.2)]

Species  Value
ReadingDateTime                
24/01/2017 19:00     NO2  120.2

In [23]:
pvt_df.groupby('Species').min(numeric_only=True)

Value
Species       
NO        -2.0
NO2        2.8
NOX        1.0
PM10      -5.6
PM2.5     -8.0

In [24]:
pvt_df[(pvt_df['Species'] =='PM10') & (pvt_df["Value"] == -5.6)]

Species  Value
ReadingDateTime                
12/09/2017 23:00    PM10   -5.6

In [30]:
pvt_df.groupby('Species').median(numeric_only=True)

Value
Species       
NO         7.1
NO2       35.3
NOX       46.9
PM10      19.5
PM2.5     11.0

In [31]:
pvt_df.groupby('Species').quantile(0.25)

Value
Species       
NO         3.1
NO2       23.8
NOX       30.3
PM10      13.9
PM2.5      7.0

7.	Dibuja un histograma para cada columna de `pvt_df`. ¿Qué contaminante tiene mayor variabilidad?

In [32]:
pvt_df.T

ReadingDateTime 01/01/2017 00:00 01/01/2017 01:00 01/01/2017 02:00  \
Species                       NO               NO               NO   
Value                        3.5              3.6              2.2   

ReadingDateTime 01/01/2017 03:00 01/01/2017 04:00 01/01/2017 05:00  \
Species                       NO               NO               NO   
Value                        2.1              3.3              2.1   

ReadingDateTime 01/01/2017 06:00 01/01/2017 07:00 01/01/2017 08:00  \
Species                       NO               NO               NO   
Value                        4.4              5.1              7.4   

ReadingDateTime 01/01/2017 09:00  ... 31/12/2017 14:00 31/12/2017 15:00  \
Species                       NO  ...            PM2.5            PM2.5   
Value                        6.5  ...              8.0              2.0   

ReadingDateTime 31/12/2017 16:00 31/12/2017 17:00 31/12/2017 18:00  \
Species                    PM2.5            PM2.5            PM2.5   
Value                        2.0              4.0              5.0   

ReadingDateTime 31/12/2017 19:00 31/12/2017 20:00 31/12/2017 21:00  \
Species                    PM2.5            PM2.5            PM2.5   
Value                       -2.0              6.0              5.0   

ReadingDateTime 31/12/2017 22:00 31/12/2017 23:00  
Species                    PM2.5            PM2.5  
Value                        5.0              6.0  

[2 rows x 43800 columns]

8.	Ejecuta el siguiente código y comenta qué hace cada línea:

In [ ]:
datetime_df = air_df.ReadingDateTime.str.split(' ',expand=True)
datetime_df.columns = ['Date','Time']
datetime_df
date_df = datetime_df.Date.str.split('/',expand=True)
date_df.columns = ['Day','Month','Year']
date_df
air_df = air_df.join(date_df).join(datetime_df.Time).drop(columns=['ReadingDateTime','Year'])
air_df = air_df.set_index(['Month','Day','Time','Species'])
air_df

9. Ejecuta la siguiente instrucción y compara su salida con la del código anterior (ejercicio 8). ¿Son iguales?

In [ ]:
air_df.unstack()

10.	Explica cuáles son las diferencias y similitudes entre el par `melt()/pivot()` y el par `stack()/unstack()`. Si tuvieras que elegir una contraparte para `melt()` entre `stack()/unstack()`, ¿cuál elegirías? ¿Por qué?